File: packages/pvm-assemblyscript/assembly/instructions/rotations.ts
import {
  OPCODE_ROT_R_32_IMM,
  OPCODE_ROT_R_32_IMM_ALT,
  OPCODE_ROT_R_64_IMM,
  OPCODE_ROT_R_64_IMM_ALT,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

export class ROT_R_64_IMMInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_ROT_R_64_IMM
  name: string = 'ROT_R_64_IMM'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValue(context.registers, registerB)

    // Ensure rotation amount is within 64-bit range
    const rotationAmount = u64(immediateX) % u64(64)
    const result = this.rotateRight64(registerValue, rotationAmount)

    this.setRegisterValue(context.registers, registerA, result)

    return new InstructionResult(-1)
  }

  rotateRight64(value: u64, amount: u64): u64 {
    // Handle negative rotation amounts
    let normalizedAmount = amount
    if (i64(amount) < i64(0)) {
      normalizedAmount = u64(64) + amount
    }

    // Normalize amount to 0-63 range
    normalizedAmount = normalizedAmount % u64(64)

    if (normalizedAmount === u64(0)) return value

    // Perform right rotation
    const mask = u64(0xffffffffffffffff)
    const rightPart = value >> normalizedAmount
    const leftPart = (value << (u64(64) - normalizedAmount)) & mask
    return (rightPart | leftPart) & mask
  }
}

export class ROT_R_64_IMM_ALTInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_ROT_R_64_IMM_ALT
  name: string = 'ROT_R_64_IMM_ALT'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    // Gray Paper: rotates immediateX by registerB
    const immediateValue = u64(immediateX)
    const registerBValue = this.getRegisterValue(context.registers, registerB)

    // Alternative rotation implementation: rotate immediate by register value
    const rotationAmount = i32(registerBValue % u64(64))
    const result = this.rotateRight64(immediateValue, u64(rotationAmount))

    this.setRegisterValue(context.registers, registerA, result)

    return new InstructionResult(-1)
  }

  rotateRight64(value: u64, amount: u64): u64 {
    // Handle negative rotation amounts
    let normalizedAmount = amount
    if (i64(amount) < i64(0)) {
      normalizedAmount = u64(64) + amount
    }

    // Normalize amount to 0-63 range
    normalizedAmount = normalizedAmount % u64(64)

    if (normalizedAmount === u64(0)) return value

    // Perform right rotation
    const mask = u64(0xffffffffffffffff)
    const rightPart = value >> normalizedAmount
    const leftPart = (value << (u64(64) - normalizedAmount)) & mask
    return (rightPart | leftPart) & mask
  }
}

export class ROT_R_32_IMMInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_ROT_R_32_IMM
  name: string = 'ROT_R_32_IMM'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const registerValue =
      this.getRegisterValue(context.registers, registerB) & u64(0xffffffff)

    // Ensure rotation amount is within 32-bit range
    const rotationAmount = i32(u64(immediateX) % u64(32))
    const result = this.rotateRight32(registerValue, rotationAmount)

    // Gray Paper: reg'_A = sext{4}{x} - sign-extend 32-bit result to 64 bits
    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }

  rotateRight32(value: u64, amount: i32): u64 {
    // Handle negative rotation amounts
    let normalizedAmount = amount
    if (amount < 0) {
      normalizedAmount = 32 + amount
    }

    // Normalize amount to 0-31 range
    normalizedAmount = normalizedAmount % 32

    if (normalizedAmount === 0) return value

    // Perform right rotation
    const mask = u64(0xffffffff)
    const rightPart = value >> u64(normalizedAmount)
    const leftPart = (value << u64(32 - normalizedAmount)) & mask
    return (rightPart | leftPart) & mask
  }
}

export class ROT_R_32_IMM_ALTInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_ROT_R_32_IMM_ALT
  name: string = 'ROT_R_32_IMM_ALT'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseTwoRegistersAndImmediate(context.operands, context.fskip)
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    // Gray Paper: rotates immediateX by registerB (32-bit)
    const immediateValue = u64(immediateX) & u64(0xffffffff)
    const registerBValue =
      this.getRegisterValue(context.registers, registerB) & u64(0xffffffff)

    // Alternative rotation implementation: rotate immediate by register value
    const rotationAmount = i32(registerBValue % u64(32))
    const result = this.rotateRight32(immediateValue, rotationAmount)

    // Gray Paper: reg'_A = sext{4}{x} - sign-extend 32-bit result to 64 bits
    this.setRegisterValueWith32BitResult(context.registers, registerA, result)

    return new InstructionResult(-1)
  }

  rotateRight32(value: u64, amount: i32): u64 {
    // Handle negative rotation amounts
    let normalizedAmount = amount
    if (amount < 0) {
      normalizedAmount = 32 + amount
    }

    // Normalize amount to 0-31 range
    normalizedAmount = normalizedAmount % 32

    if (normalizedAmount === 0) return value

    // Perform right rotation
    const mask = u64(0xffffffff)
    const rightPart = value >> u64(normalizedAmount)
    const leftPart = (value << u64(32 - normalizedAmount)) & mask
    return (rightPart | leftPart) & mask
  }
}

File: packages/pvm-rust/src/instructions/rotations.rs
//! Rotate right by immediate: ROT_R_64_IMM, ROT_R_64_IMM_ALT, ROT_R_32_IMM, ROT_R_32_IMM_ALT.

use crate::config::{
    OPCODE_ROT_R_32_IMM, OPCODE_ROT_R_32_IMM_ALT, OPCODE_ROT_R_64_IMM, OPCODE_ROT_R_64_IMM_ALT,
};
use crate::instructions::base::{
    parse_two_registers_and_immediate, sign_extend, InstructionHandler,
};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register(registers: &mut [u64; 13], index: u8, value: u64) {
    if (index as usize) < 13 {
        registers[index as usize] = value;
    }
}

fn set_register_32(registers: &mut [u64; 13], index: u8, value: u64) {
    let extended = sign_extend(value & 0xffff_ffff, 4);
    if (index as usize) < 13 {
        registers[index as usize] = extended;
    }
}

fn rot_right_64(value: u64, amount: u32) -> u64 {
    let amount = amount % 64;
    if amount == 0 {
        return value;
    }
    let mask = 0xffff_ffff_ffff_ffffu64;
    let right = value >> amount;
    let left = (value << (64 - amount)) & mask;
    right | left
}

fn rot_right_32(value: u64, amount: u32) -> u64 {
    let value = value & 0xffff_ffff;
    let amount = amount % 32;
    if amount == 0 {
        return value;
    }
    let mask = 0xffff_ffffu64;
    let right = value >> amount;
    let left = (value << (32 - amount)) & mask;
    right | left
}

// ROT_R_64_IMM: rotate reg_B right by immed_X, store in reg_A
pub struct RotR64ImmInstruction;
impl RotR64ImmInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for RotR64ImmInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ROT_R_64_IMM)
    }
    fn name(&self) -> &'static str {
        "ROT_R_64_IMM"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let value = get_register(context.registers, p.register_b);
        let amount = (p.immediate_x as u64) % 64;
        let result = rot_right_64(value, amount as u32);
        set_register(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// ROT_R_64_IMM_ALT: rotate immed_X right by reg_B, store in reg_A
pub struct RotR64ImmAltInstruction;
impl RotR64ImmAltInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for RotR64ImmAltInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ROT_R_64_IMM_ALT)
    }
    fn name(&self) -> &'static str {
        "ROT_R_64_IMM_ALT"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let value = p.immediate_x as u64;
        let amount = (get_register(context.registers, p.register_b) % 64) as u32;
        let result = rot_right_64(value, amount);
        set_register(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// ROT_R_32_IMM: rotate reg_B (32-bit) right by immed_X, store in reg_A (sext)
pub struct RotR32ImmInstruction;
impl RotR32ImmInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for RotR32ImmInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ROT_R_32_IMM)
    }
    fn name(&self) -> &'static str {
        "ROT_R_32_IMM"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let value = get_register(context.registers, p.register_b) & 0xffff_ffff;
        let amount = (p.immediate_x as u64) % 32;
        let result = rot_right_32(value, amount as u32);
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// ROT_R_32_IMM_ALT: rotate immed_X (32-bit) right by reg_B, store in reg_A (sext)
pub struct RotR32ImmAltInstruction;
impl RotR32ImmAltInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for RotR32ImmAltInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ROT_R_32_IMM_ALT)
    }
    fn name(&self) -> &'static str {
        "ROT_R_32_IMM_ALT"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_two_registers_and_immediate(context.operands, context.fskip);
        let value = (p.immediate_x as u64) & 0xffff_ffff;
        let amount = (get_register(context.registers, p.register_b) % 32) as u32;
        let result = rot_right_32(value, amount);
        set_register_32(context.registers, p.register_a, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
